deps: update V8 to 7.0

targos opened the pull request:
ETA: Oct 16th, 2018

Issues to fix:
- [x] https://github.com/nodejs/node-v8/issues/77
- [x] ~https://github.com/nodejs/node-v8/issues/78~ https://chromium-review.googlesource.com/c/v8/v8/+/1235927
- [x] `v8ustack.d` (SmartOS)
- [x] Missing postmortem metadata constants
- [x] WASM errors on PPC

targos commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/17076/
V8: https://ci.nodejs.org/job/node-test-commit-v8-linux/1671/

vsemozhetbyt commented:
Hello, [stable Array.prototype.sort()](https://twitter.com/mathias/status/1036626116654637057).

targos commented:
This version is also affected by https://github.com/nodejs/node-v8/issues/77

/cc @refack who's been working on it.

targos commented:
Pushed temporary fixes for https://github.com/nodejs/node-v8/issues/77 and https://github.com/nodejs/node-v8/issues/78.

CI: https://ci.nodejs.org/job/node-test-pull-request/17087/
V8: https://ci.nodejs.org/job/node-test-commit-v8-linux/1675/

AyushG3112 commented:
Just confirming, this will release with nodejs 11 right?

targos commented:
There are missing postmortem metadata constants:

```
01:24:58     AssertionError [ERR_ASSERTION]: Missing constants: v8dbg_class_SharedFunctionInfo__function_identifier_or_debug_info__Object,v8dbg_class_SharedFunctionInfo__script__Object
```

targos commented:
@nodejs/platform-ppc something related to WASM seems broken:

https://ci.nodejs.org/job/node-test-commit-v8-linux/1675/

targos commented:
Ping @nodejs/platform-smartos @nodejs/platform-ppc

mhdawson commented:
@john-yan Please look at the ppc issue and comment here.

john-yan commented:
Hello Michael, I am aware of the errors and trying to fix asap

john-yan commented:
Hello @targos , the ppc Atomic errors shown in https://ci.nodejs.org/job/node-test-commit-v8-linux/1675/#showFailuresLink are skipped by https://chromium-review.googlesource.com/c/v8/v8/+/1217003 as I64Atomic operations are not yet supported on the platform.

targos commented:
Thanks @john-yan, I just included the changes.

CI: https://ci.nodejs.org/job/node-test-pull-request/17231/
V8 CI: https://ci.nodejs.org/job/node-test-commit-v8-linux/1688/

targos commented:
@cjihrig maybe you can have a look at the postmortem issues?

There is a small error with `src/v8ustack.d`:

```
01:17:47 dtrace: failed to compile script src/v8ustack.d: line 402: failed to resolve V8DBG_CLASS_SHAREDFUNCTIONINFO__FUNCTION_IDENTIFIER_OR_DEBUG_INFO__OBJECT: Unknown variable name
```

There are missing postmortem metadata constants:

```
01:24:58     AssertionError [ERR_ASSERTION]: Missing constants: v8dbg_class_SharedFunctionInfo__function_identifier_or_debug_info__Object,v8dbg_class_SharedFunctionInfo__script__Object
```

cjihrig commented:
I'll take a look.

targos commented:
@cjihrig Done. Thanks a lot!

targos commented:
Only https://github.com/nodejs/node-v8/issues/78 remains. /cc @ofrobots

refack commented:
@targos I see a similar `v8ustack.d` error with 7.1, probably need a cherry-pick.
```
20:52:01   LD_LIBRARY_PATH=/home/iojs/build/workspace/node-test-commit-smartos/nodes/smartos17-64/out/Release/lib.host:/home/iojs/build/workspace/node-test-commit-smartos/nodes/smartos17-64/out/Release/lib.target:$LD_LIBRARY_PATH; export LD_LIBRARY_PATH; cd ../.; mkdir -p /home/iojs/build/workspace/node-test-commit-smartos/nodes/smartos17-64/out/Release/obj.target/node_lib/src; dtrace -64 "-I/home/iojs/build/workspace/node-test-commit-smartos/nodes/smartos17-64/out/Release/obj/gen" -Isrc -C -G -s src/v8ustack.d -o "/home/iojs/build/workspace/node-test-commit-smartos/nodes/smartos17-64/out/Release/obj.target/node_lib/src/node_dtrace_ustack.o"
20:52:01 In file included from ../src/node_internals.h:30:0,
20:52:01                  from ../src/node_dtrace.h:27,
20:52:01                  from ../src/node_dtrace.cc:22:
20:52:01 ../src/util-inl.h:43:0: warning: "BSWAP_8" redefined
20:52:01  #define BSWAP_8(x)                                                            \
20:52:01  
20:52:01 In file included from /usr/include/netinet/in.h:72:0,
20:52:01                  from /usr/include/sys/socket.h:67,
20:52:01                  from ../deps/uv/include/uv/unix.h:30,
20:52:01                  from ../deps/uv/include/uv.h:62,
20:52:01                  from ../src/node_mutex.h:7,
20:52:01                  from ../src/node_internals.h:28,
20:52:01                  from ../src/node_dtrace.h:27,
20:52:01                  from ../src/node_dtrace.cc:22:
20:52:01 /usr/include/sys/byteorder.h:107:0: note: this is the location of the previous definition
20:52:01  #define BSWAP_8(x) ((x) & 0xff)
20:52:01
....
20:52:08 dtrace: failed to compile script src/v8ustack.d: line 387: failed to resolve V8DBG_CLASS_STRING__LENGTH__SMI: Unknown variable name
20:52:08 node_dtrace_ustack.target.mk:26: recipe for target '/home/iojs/build/workspace/node-test-commit-smartos/nodes/smartos17-64/out/Release/obj.target/node_lib/src/node_dtrace_ustack.o' failed
20:52:08 make[2]: *** [/home/iojs/build/workspace/node-test-commit-smartos/nodes/smartos17-64/out/Release/obj.target/node_lib/src/node_dtrace_ustack.o] Error 1
```

https://ci.nodejs.org/job/node-test-commit-smartos/20332/nodes=smartos17-64/

targos commented:
I included both fixes from https://github.com/nodejs/node-v8/issues/78. Still hoping we can get them backported upstream.

CI: https://ci.nodejs.org/job/node-test-pull-request/17317/
V8 CI: https://ci.nodejs.org/job/node-test-commit-v8-linux/1703/

targos commented:
@refack we'll have to float a patch similar to https://github.com/nodejs/node/pull/22754/commits/d39b90b16f6de963cc11490a76165e38c0541220 on `canary-base`.

targos commented:
It looks like the update breaks https://github.com/nodejs/node/pull/22788
I'll investigate.

targos commented:
^ Is probably related to the fact that V8 now has a stable sorting algorithm in v7.0.

V8 6.9:
```
>  [100, 200, 101].sort((a, b) => a > b)
[ 100, 101, 200 ]
>  [100, 200, 101].sort((a, b) => a < b)
[ 200, 101, 100 ]
```

V8 7.0:
```
>  [100, 200, 101].sort((a, b) => a > b)
[ 100, 200, 101 ]
>  [100, 200, 101].sort((a, b) => a < b)
[ 100, 200, 101 ]
```

We could say that the comparison function is buggy (and that would be correct, because it never returns a negative value). However, it has always been working like that, so I'm afraid that many applications rely on it.
I'm going to fix `util.inspect` but we'll need to clearly document this new behavior in the blog post for Node 11.

motss commented:
This is interesting. I just tried the small snippet in latest Chrome Canary and this is my result:

```js
>  [100, 200, 101].sort((a, b) => b > a)
 (3) [100, 200, 101]

> [100, 200, 101].sort((a, b) => b - a)
(3) [200, 101, 100]
```

I always think that you can do sorting with `b - a` in `sort`.

refack commented:
> so I'm afraid that many applications rely on it.

A very limited grep on a year old gzemnid, found ~300 lines in ~50 packages.

ofrobots commented:
@hashseed what is the likelihood of timsort being reverted before 7.0 hits stable?

Either way, landing 7.0 on master is fine as long as we don't ship Node 11.0.0 before V8 7.0 hits stable.

targos commented:
Rebased on top of https://github.com/nodejs/node/pull/22992.
Updated V8 to 7.0.276.20

CI: https://ci.nodejs.org/job/node-test-pull-request/17370/
V8: https://ci.nodejs.org/job/node-test-commit-v8-linux/1710/

targos commented:
Landed in 13245dc50d..8aca934009.

Happy recompiling 😀

addaleax commented:
The removal of the deprecated versions of V8 methods breaks a *lot* of addons, and generates a large number of failing CITGM entries… Is there something we want to do about it?

refack commented:
Adding an extra `dont-land-on-v10.x` due to `array.sort` implementation change.

targos commented:
It's Semver-major. There's no risk for it to land on v10.x
